src/librustc_traits/chalk/db.rs

```diff
@@ -0,0 +1,479 @@
+use rustc_middle::traits::{ChalkRustDefId as RustDefId, ChalkRustInterner as RustInterner};
+use rustc_middle::ty::subst::{InternalSubsts, Subst, SubstsRef};
+use rustc_middle::ty::{self, AssocItemContainer, AssocKind, TyCtxt};
+
+use rustc_hir::def_id::DefId;
+
+use rustc_span::symbol::sym;
+
+use std::fmt;
+use std::sync::Arc;
+
+use crate::lowering::LowerInto;
+
+pub struct RustIrDatabase<'tcx> {
+    pub tcx: TyCtxt<'tcx>,
+    pub interner: RustInterner<'tcx>,
+}
+
+impl fmt::Debug for RustIrDatabase<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "RustIrDatabase")
+    }
+}
+
+impl<'tcx> chalk_solve::RustIrDatabase<RustInterner<'tcx>> for RustIrDatabase<'tcx> {
+    fn interner(&self) -> &RustInterner<'tcx> {
+        &self.interner
+    }
+
+    fn associated_ty_data(
+        &self,
+        assoc_type_id: chalk_ir::AssocTypeId<RustInterner<'tcx>>,
+    ) -> Arc<chalk_rust_ir::AssociatedTyDatum<RustInterner<'tcx>>> {
+        let def_id = match assoc_type_id.0 {
+            RustDefId::AssocTy(def_id) => def_id,
+            _ => bug!("Did not use `AssocTy` variant when expecting associated type."),
+        };
+        let assoc_item = self.tcx.associated_item(def_id);
+        let trait_def_id = match assoc_item.container {
+            AssocItemContainer::TraitContainer(def_id) => def_id,
+            _ => unimplemented!("Not possible??"),
+        };
+        match assoc_item.kind {
+            AssocKind::Type => {}
+            _ => unimplemented!("Not possible??"),
+        }
+        let trait_bound_vars = bound_vars_for_item(self.tcx, trait_def_id);
+        let binders = binders_for(&self.interner, trait_bound_vars);
+
+        Arc::new(chalk_rust_ir::AssociatedTyDatum {
+            trait_id: chalk_ir::TraitId(RustDefId::Trait(trait_def_id)),
+            id: assoc_type_id,
+            name: (),
+            binders: chalk_ir::Binders::new(
+                binders,
+                chalk_rust_ir::AssociatedTyDatumBound { bounds: vec![], where_clauses: vec![] },
```

Comment:
this is clearly not right..?

Reply:
Indeed. I added some code here. But I don't think it's exactly right. (Namely, I'm not entirely sure where to find the predicates for bounds vs. where clauses for associated types)